reactor-core/src/main/java/reactor/core/publisher/FluxSwitchOnFirst.java

```diff
@@ -523,58 +559,141 @@ public void onError(Throwable throwable) {
                 parent.cancel();
             }
 
-            inner.onError(throwable);
+            delegate.onError(throwable);
         }
 
         @Override
         public void onComplete() {
-            if (!parent.done) {
+            if (!parent.done && cancelSourceOnComplete) {
                 parent.cancel();
             }
 
-            inner.onComplete();
+            delegate.onComplete();
+        }
+
+        @Override
+        public void request(long n) {
+            long r = this.requested;
+
+            if (r > STATE_REQUESTED) {
+                long u;
+                for (;;) {
+                    if (r == Long.MAX_VALUE) {
+                        return;
+                    }
+                    u = Operators.addCap(r, n);
+                    if (REQUESTED.compareAndSet(this, r, u)) {
+                        return;
+                    }
+                    else {
+                        r = requested;
+
+                        if (r < 0) {
+                            break;
+                        }
+                    }
+                }
+            }
+
+            if (r == STATE_CANCELLED) {
+                return;
+            }
+
+            s.request(n);
```

Comment:
this is only done when `this.requested == STATE_REQUESTED`, ie in `tryRequest`, in turn only invoked in `onSubscribe`, correct?

Reply:
correct, since there is a guard for that at line 578 I'm not trying to double-check that again here. it is assumed if the execution got there than the state is requested

Reply:
what I meant is "what if the state was NOT requested, but a second batch of request (eg. this operator is called with `request(3)` then `request(4)`). wouldn't the second request call be "lost"?

Reply:
It is not boing to be lost.

State 1 [r: 0]

request(3) -> State 2 [r: 3]

State 2 [r: 3] race with [onSubscribe]
Case 1: 
	tryRequest( getAndSet(-1) ) Happens Before request( REQUESTED.addAndGet(4) )

Result: 
	request(3) followed by
    request(4) with a happens-before guarantee

Case 2: 
   request ( REQUESTED.addAndGet(4)) happens before tryRequest( getAndSet(-1) )

Result: 
    request(7)